crates/telio-starcast/src/transport.rs

```diff
@@ -0,0 +1,243 @@
+            })?;
+        let transport_socket = Arc::new(transport_socket);
+
+        Ok(Self {
+            task: Task::start(State {
+                vpeer_ip,
+                transport_socket,
+                packet_chan,
+                recv_buffer: vec![0; MAX_PACKET_SIZE],
+                nat_map: RwLock::new(HashMap::new()),
+                peers: Vec::new(),
+                multicast_ips: Vec::new(),
+            }),
+        })
+    }
+
+    pub async fn configure(&self, config: Config) -> Result<(), Error> {
+        task_exec!(&self.task, async move |state| {
+            state.configure(config);
+            Ok(())
+        })
+        .await?;
+        Ok(())
+    }
+
+    pub async fn stop(self) {
+        let _ = self.task.stop().await.resume_unwind();
+    }
+}
+
+struct State {
+    vpeer_ip: IpAddr,
+    transport_socket: Arc<UdpSocket>,
+    packet_chan: Chan<Vec<u8>>,
+    recv_buffer: Vec<u8>,
+    nat_map: RwLock<HashMap<u16, SocketAddr>>,
+    peers: Vec<IpAddr>,
+    multicast_ips: Vec<IpNetwork>,
+}
+
+impl State {
+    fn configure(&mut self, config: Config) {
+        self.peers = config.peers;
+        self.multicast_ips = config.multicast_ips;
+    }
+
+    async fn handle_local_multicast_packet(&mut self, packet: Vec<u8>) {
+        let fanout_res = join_all(self.peers.iter().map(|peer| {
+            self.transport_socket
+                .send_to(&packet, SocketAddr::new(*peer, MULTICAST_TRANSPORT_PORT))
+        }))
+        .await;
+        let fail_count = fanout_res.iter().filter(|res| res.is_err()).count();
+        if fail_count > 0 {
+            telio_log_debug!(
+                "Failed to fanout local multicast packet to {fail_count}/{} peers",
+                fanout_res.len()
+            );
+        }
+    }
+
+    async fn handle_mapped_unicast_packet(&mut self, mut packet: Vec<u8>) {
+        let natted_port = u16::from_be_bytes([
+            *(packet.get(IPV4_HEADER + UDP_DEST_OFFSET).unwrap_or(&0)),
+            *(packet.get(IPV4_HEADER + UDP_DEST_OFFSET + 1).unwrap_or(&0)),
+        ]);
+        let (dst_ip, dst_port) = if let Some((IpAddr::V4(addr), dst_port)) = self
+            .nat_map
+            .read()
+            .await
+            .get(&natted_port)
+            .map(|sock_addr| (sock_addr.ip(), sock_addr.port()))
+        {
+            (addr, dst_port)
+        } else {
+            return;
+        };
+        if let Some(mut ip_packet) = MutableIpv4Packet::new(&mut packet) {
+            ip_packet.set_destination(dst_ip);
+            ip_packet.set_checksum(0);
+            ip_packet.set_checksum(checksum(&ip_packet.to_immutable()));
+        } else {
+            return;
+        };
+        if let Some(mut udp_packet) = packet
+            .get_mut(IPV4_HEADER..)
+            .and_then(MutableUdpPacket::new)
+        {
+            udp_packet.set_destination(dst_port);
+            // Ignoring L4 checksum as it is already encapsulated in L3/L4 header
```

Comment:
This is the fun part, that you actualy need to calculate both ip and udp :D. But we can wait for @jjanowsk nat

Reply:
That's why I haven't touched these packet rewrite sections since the PR opened, because it will be covered by the natting code anyways :smile: